Libraries/LibCore/CGzip.cpp

```diff
@@ -0,0 +1,120 @@
+            ASSERT_NOT_REACHED();
+            return (u8)0;
+        }
+        dbg() << "read_byte: " << String::format("%x", data[current]);
+        return data[current++];
+    };
+
+    dbg() << "get_gzip_payload: Skipping over gzip header.";
+
+    // Magic Header
+    if (read_byte() != 0x1F || read_byte() != 0x8B) {
+        dbg() << "get_gzip_payload: Wrong magic number.";
+        return Optional<AK::ByteBuffer>();
+    }
+
+    // Compression method
+    auto method = read_byte();
+    if (method != 8) {
+        dbg() << "get_gzip_payload: Wrong compression method = " << method;
+        return Optional<AK::ByteBuffer>();
+    }
+
+    u8 flags = read_byte();
+
+    // Timestamp, Extra flags, OS
+    current += 6;
+
+    // FEXTRA
+    if (flags & 4) {
+        u16 length = read_byte() & read_byte() << 8;
+        dbg() << "get_gzip_payload: Header has FEXTRA flag set. Length = " << length;
+        current += length;
+    }
+
+    // FNAME
+    if (flags & 8) {
+        dbg() << "get_gzip_payload: Header has FNAME flag set.";
+        while (read_byte() != '\0')
+            ;
+    }
+
+    // FCOMMENT
+    if (flags & 16) {
+        dbg() << "get_gzip_payload: Header has FCOMMENT flag set.";
+        while (read_byte() != '\0')
+            ;
+    }
+
+    // FHCRC
+    if (flags & 2) {
+        dbg() << "get_gzip_payload: Header has FHCRC flag set.";
+        current += 2;
+    }
+
+    auto new_size = data.size() - current;
+    dbg() << "get_gzip_payload: Returning slice from " << current << " with size " << new_size;
+    return data.slice(current, new_size);
+}
+
+Optional<ByteBuffer> Gzip::decompress(const ByteBuffer& data)
+{
+    ASSERT(is_compressed(data));
+
+    dbg() << "Gzip::decompress: Decompressing gzip compressed data. Size = " << data.size();
+    auto optional_payload = get_gzip_payload(data);
+    if (!optional_payload.has_value()) {
+        return Optional<AK::ByteBuffer>();
+    }
+
+    auto source = optional_payload.value();
+    unsigned long source_len = source.size();
+    auto destination = ByteBuffer::create_uninitialized(1024);
+    while (true) {
+        unsigned long destination_len = destination.size();
+        // FIXME: dbg() cannot take ulong?
+        dbg() << "Gzip::decompress: Calling puff()\n"
+              << "  destination_data = " << destination.data() << "\n"
+              << "  destination_len = " << (int)destination_len << "\n"
+              << "  source_data = " << source.data() << "\n"
+              << "  source_len = " << (int)source_len;
+
+        auto puff_ret = puff(
+            destination.data(), &destination_len,
+            source.data(), &source_len);
+
+        if (puff_ret == 0) {
+            dbg() << "Gzip::decompress: Decompression success.";
+            break;
+        } else if (puff_ret == 1) {
+            dbg() << "Gzip::decompress: Output buffer exhausted. Growing.";
+            destination.grow(destination.size() * 2);
```

Comment:
This is sad. We have to try decompressing from scratch again and again, until we have a buffer that's big enough?
It's okay for a first implementation but we can definitely improve this. :)

Reply:
Yeah you're right. zlib handled this much better because I could inflate() 1024 byte chunks one after the other. But I don't think puff can do this.

I have added a FIXME comment.